Tidy & reword Secret concept (#30984)
kubernetes/website: content/en/docs/concepts/configuration/secret.md

[bootstrap token Secrets](#bootstrap-token-secrets) are a mechanism to
help automate node registration.

### Alternatives to Secrets

Rather than using a Secret to protect confidential data, you can pick from alternatives.

Here are some of your options:

- if your cloud-native component needs to authenticate to another application that you
  know is running within the same Kubernetes cluster, you can use a

>   [service account](/docs/reference/access-authn-authz/authentication/#service-account-tokens)

Reviewer: ```suggestion
  [ServiceAccount](/docs/reference/access-authn-authz/authentication/#service-account-tokens)
```
nit: upper camel case?

Author: Sure.

`stringData` field, the value specified in the `stringData` field takes
precedence.

## Types of Secret {#secret-types}

When creating a Secret, you can specify its type using the `type` field of
a Secret resource, or certain equivalent `kubectl` command line flags (if available).
The `type` of a Secret is used to facilitate programmatic handling of different
kinds of confidential data.

Kubernetes provides several builtin types for some common usage scenarios.
These types vary in terms of the validations performed and the constraints
Kubernetes imposes on them.

| Builtin Type | Usage |
|--------------|-------|
| `Opaque`     |  arbitrary user-defined data |
| `kubernetes.io/service-account-token` | service account token |
| `kubernetes.io/dockercfg` | serialized `~/.dockercfg` file |
| `kubernetes.io/dockerconfigjson` | serialized `~/.docker/config.json` file |
| `kubernetes.io/basic-auth` | credentials for basic authentication |
| `kubernetes.io/ssh-auth` | credentials for SSH authentication |
| `kubernetes.io/tls` | data for a TLS client or server |

> | `bootstrap.kubernetes.io/token` | bootstrap token data |

Author: This is retained; I move it to later in the page.

to be used by a container in a Pod. Secrets can also be used by other parts of the
system, without being directly exposed to the Pod. For example, Secrets can hold
credentials that other parts of the system should use to interact with external
systems on your behalf.

Secret volume sources are validated to ensure that the specified object
reference actually points to an object of type Secret. Therefore, a Secret
needs to be created before any Pods that depend on it.  

If the Secret cannot be fetched (perhaps because it does not exist, or
due to a temporary lack of connection to the API server) the kubelet
periodically retries running that Pod. The kubelet also reports an Event
for that Pod, including details of the problem fetching the Secret.

#### Optional Secrets {#restriction-secret-must-exist}

When you define a container environment variable based on a Secret,
you can mark it as _optional_. The default is for the Secret to be
required.

None of a Pod's containers will start until all non-optional Secrets are
available.

If a Pod references a specific key in a Secret and that Secret does exist, but
is missing the named key, the Pod fails during startup.

> ### Using Secrets as files from a Pod

Author: ```suggestion
### Using Secrets as files from a Pod {#using-secrets-as-files-from-a-pod}
```

> To consume a Secret in a volume in a Pod:

Author: I'm fine to retain the existing wording (a later PR could update it), but if we want to change it, how about:

```suggestion
If you want to access data from a Secret in a Pod, one way to do that is to
have Kubernetes make the value of that Secret be available as a file inside
the filesystem of one or more of the Pod's containers.

To configure that, you:
```

> To consume a Secret in a volume in a Pod:

Reviewer 2: I like that. It's more precise.

  `/etc/foo/my-group/my-username` instead of at `/etc/foo/username`.
* the `password` key from that Secret object is not projected.

If `.spec.volumes[].secret.items` is used, only keys specified in `items` are projected.
To consume all keys from the Secret, all of them must be listed in the `items` field.

If you list keys explicitly, then all listed keys must exist in the corresponding Secret.
Otherwise, the volume is not created.

#### Secret files permissions

You can set the POSIX file access permission bits for a single Secret key.
If you don't specify any permissions, `0644` is used by default.
You can also set a default mode for the entire Secret volume and override per key if needed.

For example, you can specify a default mode like this:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: mypod
spec:
  containers:
  - name: mypod
    image: redis
    volumeMounts:
    - name: foo
      mountPath: "/etc/foo"
  volumes:
  - name: foo
    secret:
      secretName: mysecret
      defaultMode: 0400
```

The secret is mounted on `/etc/foo`; all the files created by the
secret volume mount have permission `0400`.

{{< note >}}

> If you're defining a Pod or a Pod template using JSON, beware that the JSON

Author: Not convinced here. Is “beware” unidiomatic?

> If you're defining a Pod or a Pod template using JSON, beware that the JSON

Reviewer 2: I don't know that we would use the words differently on either side of the pond, but I leave it to you. I just thought that "beware" implies some danger while "be aware" is just something that would be helpful to know. If you think this facts represents danger, then beware is fine.

> If you're defining a Pod or a Pod template using JSON, beware that the JSON

Author: There is some danger here. People might specify the wrong `defaultMode` and leak a secret, because they thought that a number starting with `0` would be treated as octal.

> See [Add ImagePullSecrets to a service account](/docs/tasks/configure-pod-container/configure-service-account/#add-imagepullsecrets-to-a-service-account)

Reviewer 2: The link to "Add ImagePullSecrets to a service account" is repeated on about 20 lines up from here. Just want to make sure you meant to have it in both places.

 for a detailed explanation of that process.

### Using Secrets with static Pods {#restriction-static-pod}

You cannot use ConfigMaps or Secrets with
{{< glossary_tooltip text="static Pods" term_id="static-pod" >}}.

## Use cases

> ### Use case: As container environment variables

Reviewer 2: Consider adding text between the level 2 heading that the first level 3 heading to introduce the use cases below.

Author: I'm very happy to save that for another PR - this one is making substantial enough changes that I don't see the need to cover every shortcoming we can spot.

Reviewer 2: Not a problem. You have already done a ton of work here.

The `kubernetes.io/dockercfg` type is reserved to store a serialized
`~/.dockercfg` which is the legacy format for configuring Docker command line.
When using this Secret type, you have to ensure the Secret `data` field
contains a `.dockercfg` key whose value is content of a `~/.dockercfg` file
encoded in the base64 format.

The `kubernetes.io/dockerconfigjson` type is designed for storing a serialized
JSON that follows the same format rules as the `~/.docker/config.json` file
which is a new format for `~/.dockercfg`.
When using this Secret type, the `data` field of the Secret object must
contain a `.dockerconfigjson` key, in which the content for the
`~/.docker/config.json` file is provided as a base64 encoded string.

Below is an example for a `kubernetes.io/dockercfg` type of Secret:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: secret-dockercfg
type: kubernetes.io/dockercfg
data:
  .dockercfg: |
    "<base64 encoded ~/.dockercfg file>"
```

{{< note >}}
If you do not want to perform the base64 encoding, you can choose to use the
`stringData` field instead.
{{< /note >}}

When you create these types of Secrets using a manifest, the API

> server checks whether the expected key does exists in the `data` field, and

Reviewer 2: ```suggestion
server checks whether the expected key exists in the `data` field, and
```

server doesn't validate if the JSON actually is a Docker config file.

When you do not have a Docker config file, or you want to use `kubectl`
to create a Docker registry Secret, you can do:

```shell
kubectl create secret docker-registry secret-tiger-docker \
  --docker-username=tiger \
  --docker-password=pass1234 \
  --docker-email=tiger@acme.example
```

That command creates a Secret of type `kubernetes.io/dockerconfigjson`.
If you dump the `.data.dockercfgjson` field from that new Secret and then
decode it from base64:

```shell
kubectl get secret secret-tiger-docker -o jsonpath='{.data.*}' | base64 -d
```

then the output is equivalent to this JSON document (which is also a valid
Docker configuration file):

```json
{
  "auths": {
    "https://index.docker.io/v1/": {
      "username": "tiger",
      "password": "pass1234",
      "email": "tiger@acme.example",
      "auth": "dGlnZXI6cGFzczEyMzQ="
    }
  }
}
```

### Basic authentication Secret

The `kubernetes.io/basic-auth` type is provided for storing credentials needed
for basic authentication. When using this Secret type, the `data` field of the
Secret must contain one of the following two keys:

> - `username`: the user name for authentication;

Reviewer 2: ```suggestion
- `username`: the user name for authentication
```

> - `password`: the password or token for authentication.

Reviewer 2: ```suggestion
- `password`: the password or token for authentication
```

Both values for the above two keys are base64 encoded strings. You can, of
course, provide the clear text content using the `stringData` for Secret
creation.

The following manifest is an example of a basic authentication Secret:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: secret-basic-auth
type: kubernetes.io/basic-auth
stringData:
  username: admin      # required field for kubernetes.io/basic-auth
  password: t0p-Secret # required field for kubernetes.io/basic-auth
```

The basic authentication Secret type is provided only for convenience.

> You can create an `Opaque` for credentials used for basic authentication.

Reviewer 2: ```suggestion
You can create an `Opaque` type for credentials used for basic authentication.
```

> However, using a define Secret type helps unify the formats of your credentials.

Reviewer 2: I don't understand what a "define Secret type" refers to.

Author: ```suggestion
However, using the defined and public Secret type (`kubernetes.io/basic-auth`) helps other
people to understand the purpose of your Secret, and sets a convention for what key names
to expect.
```

The Kubernetes API verifies that the required keys are set for a Secret
of this type.

### SSH authentication secrets

The builtin type `kubernetes.io/ssh-auth` is provided for storing data used in
SSH authentication. When using this Secret type, you will have to specify a
`ssh-privatekey` key-value pair in the `data` (or `stringData`) field
as the SSH credential to use.

The following manifest is an example of a Secret used for SSH public/private
key authentication:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: secret-ssh-auth
type: kubernetes.io/ssh-auth
data:
  # the data is abbreviated in this example
  ssh-privatekey: |
     MIIEpQIBAAKCAQEAulqb/Y ...
```

The SSH authentication Secret type is provided only for user's convenience.

> You can create an `Opaque` for credentials used for SSH authentication.

Reviewer 2: Should this say "Opaque type" or "Opaque Secret"?

> You can create an `Opaque` for credentials used for SSH authentication.
> However, using the builtin Secret type helps unify the formats of your credentials

Author: ```suggestion
You could instead create an `Opaque` type Secret for credentials used for SSH authentication.
However, using the defined and public Secret type (`kubernetes.io/ssh-auth`) helps other
people to understand the purpose of your Secret, and sets a convention for what key names
to expect.
```

```shell
kubectl create secret tls my-tls-secret \
  --cert=path/to/cert/file \
  --key=path/to/key/file
```

The public/private key pair must exist before hand. The public key certificate
for `--cert` must be DER format as per
[Section 5.1 of RFC 7468](https://datatracker.ietf.org/doc/html/rfc7468#section-5.1),
and must match the given private key for `--key` (PKCS #8 in DER format;
[Section 11 of RFC 7468](https://datatracker.ietf.org/doc/html/rfc7468#section-11)).

{{< note >}}
A kubernetes.io/tls Secret stores the Base64-encoded DER data for keys and
certificates. If you're familiar with PEM format for private keys and for certificates,
the base64 data are the same as that format except that you omit
the initial and the last lines that are used in PEM.

For example, for a certificate, you do **not** include `--------BEGIN CERTIFICATE-----`
and `-------END CERTIFICATE----`.
{{< /note >}}

### Bootstrap token Secrets

A bootstrap token Secret can be created by explicitly specifying the Secret
`type` to `bootstrap.kubernetes.io/token`. This type of Secret is designed for
tokens used during the node bootstrap process. It stores tokens used to sign

> well known ConfigMaps.

Reviewer 2: ```suggestion
well-known ConfigMaps.
```

```yaml
apiVersion: v1
kind: Secret
metadata:
  # Note how the Secret is named
  name: bootstrap-token-5emitj
  # A bootstrap token Secret usually resides in the kube-system namespace
  namespace: kube-system
type: bootstrap.kubernetes.io/token
stringData:
  auth-extra-groups: "system:bootstrappers:kubeadm:default-node-token"
  expiration: "2020-09-13T04:39:10Z"
  # This token ID is used in the name
  token-id: "5emitj"
  token-secret: "kq4gihvszzgn1p0r"
  # This token can be used for authentication
  usage-bootstrap-authentication: "true"
  # and it can be used for signing
  usage-bootstrap-signing: "true"
```

## Immutable Secrets {#secret-immutable}

{{< feature-state for_k8s_version="v1.21" state="stable" >}}

Kubernetes lets you mark specific Secrets (and ConfigMaps) as _immutable_.
Preventing changes to the data of an existing Secret has the following benefits:

- protects you from accidental (or unwanted) updates that could cause applications outages
- (for clusters that extensively use Secrets - at least tens of thousands of unique Secret
  to Pod mounts), switching to immutable Secrets improves the performance of your cluster

>   by significantly reducing load on kube-apiserverC. The kubelet does not need to maintain

Reviewer 2: ```suggestion
  by significantly reducing load on kube-apiserver. The kubelet does not need to maintain
```
Extraneous letter C?

  a [watch] on any Secrets that are marked as immutable.

### Marking a Secret as immutable {#secret-immutable-create}

You can create an immutable Secret by setting the `immutable` field to `true`. For example,
```yaml
apiVersion: v1
kind: Secret
metadata:
  ...
data:
  ...
immutable: true
```

You can also update any existing mutable Secret to make it immutable.

{{< note >}}
Once a Secret or ConfigMap is marked as immutable, it is _not_ possible to revert this change
nor to mutate the contents of the `data` field. You can only delete and recreate the Secret.
Existing Pods maintain a mount point to the deleted Secret - it is recommended to recreate
these pods.
{{< /note >}}

## Information security for Secrets

Although ConfigMap and Secret work similarly, Kubernetes applies some additional
protection for Secret objects.

Secrets often hold values that span a spectrum of importance, many of which can
cause escalations within Kubernetes (e.g. service account tokens) and to

> external systems. Even if an individual app can reason about the power of the

Reviewer 2: I don't understand "can reason about the power of the Secrets..." Maybe it's fine. I just don't understand.

Author: Let's address this in a separate PR.

Reviewer 2: Okay.
